src/XRP/shared/rippled-json-rpc-schema.ts

```diff
@@ -18,19 +18,24 @@ interface AccountLinesResponse {
   }
 }
 
-// interface AccountLinesErrorJson {
-//   result: {
-//     error: string
-//     error_code: number
-//     error_message: string
-//     request: {
-//       account: string
-//       command: string
-//       ledger_index: string
-//     }
-//     status: string
-//   }
-// }
+interface CurrencyValuePair {
+  currency: string
+  value: string
+}
+
+interface GatewayBalancesResponse {
+  result: {
+    error?: string
+    account?: string
+    assets?: { [account: string]: CurrencyValuePair[] }
+    balances?: { [account: string]: CurrencyValuePair[] }
+    ledger_hash?: string
+    ledger_index?: number
+    obligations?: { [currencyCode: string]: string }
+    status: string
+    validated?: boolean
```

Comment:
lol this is an intense type. is there some mutual exclusion between these types? ie, is it really the case that there can be between 1 and 9 different keys? or will some of these appear when a request is successful and not show up when a request is not successful?

Reply:
got it, how about

```
BaseGatewayBalancesResult {
  status: string
}

interface GatewayBalancesErrorResult extends BaseGatewayBalancesResult {
  error: string
}

interface GatewayBalancesSuccessResult extends BaseGatewayBalancesResult {
  account: string
  // a whole lotta optional fields
}

interface GatewayBalancesResponse {
  result: GatewayBalancesSuccessResult | GatewayBalancesErrorResult
}
```

interestingly, this will make it so that typescript is able to differentiate between a `GatewayBalancesErrorResult` and a `GatewayBalancesSuccessResult` by checking if `error` is there. nice.